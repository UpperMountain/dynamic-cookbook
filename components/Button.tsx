import React from "react";
import { View, ViewProps, Text, TouchableOpacity } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";

export interface Props {
  // Outer container style overrides
  style?: ViewProps["style"];

  // What to put inside the button
  children?: string | React.ReactNode;

  // Button height
  size?: number;

  // Constrain the button's width
  width?: number;

  // Color of the button's background
  color?: string;

  // Icon element from @expo/vector-icons, default `MaterialIcons`
  iconTag?: React.ElementType;

  // Icon name to render, null for no icon but still a circle
  iconName?: string | null;

  // Color of the icon to render
  iconColor?: string;

  // Distance from edge to icon
  iconPad?: number;

  // Press handler
  onPress?: () => void;
}

export default function Button({
  children,
  style,
  size = 35,
  width,
  iconName,
  iconTag: IconTag = MaterialIcons,
  iconColor = "gray",
  iconPad = 5,
  color = "#eaeaea",
  onPress
}: Props) {
  // Don't bother with the TouchableOpacity if you can't press the button
  const PressContainer: React.ElementType =
    typeof onPress === "undefined" ? View : TouchableOpacity;

  const hasIcon = typeof iconName !== "undefined";

  const circle = hasIcon && !children;

  const inner =
    typeof children === "string" ? (
      <Text
        style={{
          flex: 1,
          fontSize: size * 0.4,
          textAlign: hasIcon ? "left" : "center",
          paddingLeft: hasIcon ? iconPad : 0
        }}
      >
        {children}
      </Text>
    ) : (
      children
    );

  return (
    <PressContainer
      style={[
        {
          backgroundColor: color,
          flexDirection: "row",
          alignItems: "center",
          justifyContent: "center",
          height: size,
          width: circle ? size : width,
          paddingLeft: iconPad,
          paddingRight: iconPad,
          borderRadius: size / 2
        },
        style
      ]}
      onPress={onPress}
    >
      {iconName != null && (
        <IconTag name={iconName} size={size - iconPad * 2} color={iconColor} />
      )}
      {inner}
    </PressContainer>
  );
}
