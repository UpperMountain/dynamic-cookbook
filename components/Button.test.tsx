import React from "react";
import Button from "./Button";
import renderer from "react-test-renderer";

it("should render without exploding", () => {
  renderer.create(<Button iconName="home">Home test</Button>).unmount();
});

it("should pass snapshot tests", () => {
  const cmp = renderer.create(
    <>
      <Button iconName="home">Home test</Button>
      <Button iconName="home" />
      <Button iconName="home" onPress={jest.fn()} />
    </>
  );
  expect(cmp.toJSON()).toMatchSnapshot();
  cmp.unmount();
});
