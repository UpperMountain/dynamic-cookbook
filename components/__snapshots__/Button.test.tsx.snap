// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass snapshot tests 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#eaeaea",
          "borderRadius": 17.5,
          "flexDirection": "row",
          "height": 35,
          "justifyContent": "center",
          "paddingLeft": 5,
          "paddingRight": 5,
          "width": undefined,
        },
        undefined,
      ]
    }
  >
    <Text />
    <Text
      style={
        Object {
          "flex": 1,
          "fontSize": 14,
          "paddingLeft": 5,
          "textAlign": "left",
        }
      }
    >
      Home test
    </Text>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#eaeaea",
          "borderRadius": 17.5,
          "flexDirection": "row",
          "height": 35,
          "justifyContent": "center",
          "paddingLeft": 5,
          "paddingRight": 5,
          "width": 35,
        },
        undefined,
      ]
    }
  >
    <Text />
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#eaeaea",
        "borderRadius": 17.5,
        "flexDirection": "row",
        "height": 35,
        "justifyContent": "center",
        "opacity": 1,
        "paddingLeft": 5,
        "paddingRight": 5,
        "width": 35,
      }
    }
  >
    <Text />
  </View>,
]
`;
